### DISCUSSION TITLE
Cancel a workflow without considering it failed?

### DISCUSSION THREAD

**Discussion by jasonkuhrt** [community/community#27174] [Actions]
<p>How can a workflow be canceld but not considered a failure?</p>
<p>The use-case is that we have a release system on every trunk commit but some types of commit should not result in a release (<code>chore</code> type commits).</p>
<p>Example here <a href="https://github.com/graphql-nexus/nexus-future/runs/503474761" rel="nofollow noopener">https://github.com/graphql-nexus/nexus-future/runs/503474761</a>.</p>
<p>We use the GitHub API to script this right now. But canceling a workflow marks the workflow as being in a failing state, which is untrue.</p>
<p>How can we short-circuit a workflow without it being considered a failure?</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/jasonkuhrt">@jasonkuhrt</a> ,</p>
<p>It’s not supported, your workflow is cancelled from Rest API command, github will treat it as a failure one.</p>
<p>Typically we will use ‘if’ expression to skip some jobs/steps while not cancelling the whole workflow. </p>
<p>Code sample as below:</p>
<pre><code>- name: set output
        run: echo ::set-output name=ifpublish::$(jq -r '.kind' &lt;&lt;&lt; $result)
        id: check
      - name: skip did not publish commits
        if: steps.check.outputs.ifpublish != 'did_not_publish'
        run: |
          publish steps....
</code></pre>
<p>Hope it helps!</p>

**Reply by asiegel812** [reply]
I am asking you to cancel my subscription. There’s nowhere to cancel it. I have contacted my bank so they are aware. Are you guys make it impossible to cancel I am counseling. Thank you for your time.

**Comment by jasonkuhrt** [comment]
<p>Thanks <a class="mention" href="/weide-zhou">@weide-zhou</a> that looks like it will solve the issue. I’ll give it a shot.</p>

**Comment by jasonkuhrt** [comment]
<p>But, how do I put an if condition on a job based on the step output of a previous job in the workflow?</p>

**Comment by ThisaruGuruge** [comment]
<p>Having a similar issue here.</p>
<p>Our use case is this:</p>
<p>We have a parent repository, and whenever there’s a push on the parent repo, we have a set of child repos, which should be built incrementally.</p>
<p>This incremental build is handled by a workflow. But when we have multiple commits (within a short time period) on the parent repo, there will be many incremental build workflow runs are triggered. Therefore, we wanted to cancel any already running workflow runs before running a new one. But this marks the cancelled workflow as a failed one, which isn’t the case here. Is there a workaround for this?</p>

**Comment by cgr71ii** [comment]
Hi! Is there a way to avoid to put `if` in all the jobs? Too much code duplicated...

**Comment by ppaulweber** [comment]
@weide-zhou I would agree with @cgr71ii ... this introduces a lot of code duplication and if you miss in a step refinement of the GHA workflow the `if` condition you end up with a half broken execution.

If there is a `gh` CLI/REST cancellation functionality, why is there not one for gracefully stopping a GHA workflow?

**Comment by shakefu** [comment]
It is possible to cancel via the GH API, but you create a "cancel request", which means you need to hold/wait the action (sometimes up to 30s or more) to prevent further steps from progressing, while polling the API to see if it has canceled. There's a couple OSS actions which implement this behavior, but it's sub-optimal at best. 

Unfortunately there is no "reserved" exit code for processes to indicate signals (e.g. `exit 127` for "Cancel" instead of "Error"), and there's no reserved namespace in `$GITHUB_OUTPUT` that could be used, and the Job process itself doesn't have permissions to signal the parent Runner process directly.

Someone would have to implement a `$GITHUB_CANCEL` or similar semaphore file which could be written to by the Step process and picked up at the end of the Step by the Runner to indicate it should Cancel the workflow via the API (same behavior as the OSS Actions), or do an in-flight modification of the remaining Job metadata to effectively add `if: false` to skip the rest of the Workflow.

Anyway.

**Comment by antitoxic** [comment]
Is there a workaround for cancelling a workflow without causing it to fail? Given that the workflow fails when it's cancelled via GitHub API, this will trigger all sorts of integrations to notify devs about a problem while there isn't one.

Putting conditional `if:` for all steps is not an option since some actions that are already executed have a `post`-workflow script that will still get executed if you just skip steps. 

We need an option to make cancellation non-failing and simply marked the workflow as `cancelled()`.

**Reply by ppaulweber** [reply]
Completely agree with @antitoxic on this and as I already asked and stated couple of months ago why there isn't a `gh` CLI/REST cancellation functionality that gracefully stops a GHA workflow?

**Comment by Yury-Fridlyand** [comment]
We have CI with concurrency group configured. It cancels all CI on a branch if a new commit pushed.
All these cancelled workflows are reported as failed in email reports. There is no way to bypass that with `if` statement.
I'll be happy to have a way to cancel without marking as failed.

**Comment by gabe-mercury** [comment]
Any updates/progress on this issues?

It's creating a lot of confusion on our dev teams who are falsely identifying commits as having failed status checks when really it was a skipped deployment.

**Comment by Gabriella439** [comment]
For the use case that @Yury-Fridlyand mentioned, our company open sourced a GitHub action you can use to work around this problem (where workflows with concurrency pollute the commit timeline with red ❌s from cancelled builds):

https://github.com/MercuryTechnologies/delete-cancelled-runs

… but it would be really nice if there were a GitHub-native solution to this problem.  This GitHub action of ours is more of a hack than a proper solution.

**Comment by marcaddeo** [comment]
Would also like to see a proper solution for this implemented.
